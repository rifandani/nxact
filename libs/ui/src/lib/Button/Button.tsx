import { MouseEventHandler, PropsWithChildren } from 'react';
import '../../index.css';

export interface ButtonProps extends PropsWithChildren<HTMLButtonElement> {
  onClick?: () => MouseEventHandler<HTMLButtonElement>;
}

export function Button({ onClick, children }: ButtonProps) {
  return (
    <button
      className="inline-block px-6 py-3 bg-blue-500 text-white font-medium text-xs leading-tight uppercase rounded shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg focus:outline-none focus:ring-0 active:bg-blue-800 active:shadow-lg transition duration-150 ease-in-out"
      onClick={onClick}
    >
      {children}
    </button>
  );
}

export default Button;
