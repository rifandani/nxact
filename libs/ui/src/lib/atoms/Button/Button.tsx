import { ColorVariant, Size } from 'libs/ui/src/types/common.type';
import { cn } from 'libs/ui/src/utils/cn.util';
import React, { cloneElement } from 'react';
import '../../../index.css';

// #region INTERFACES
type ButtonType = 'solid' | 'outlined' | 'text';

type IconProps = React.ComponentProps<'div'> & {
  size: Size;
};

type BaseButtonProps = React.ComponentProps<'button'> & {
  size: Size;
  buttonType: ButtonType;
  variant: ColorVariant;
  rounded: boolean;
};

type ButtonProps = React.ComponentProps<'button'> & {
  testId?: string;
  size?: Size;
  buttonType?: ButtonType;
  variant?: ColorVariant;
  rounded?: boolean;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
};
// #endregion

const buttonSizeClasses = {
  small: 'py-1.5 text-sm rounded text-red-500',
  medium: 'py-[9px] text-base rounded-md',
  large: 'py-3 text-lg rounded-md',
};
const buttonTypeClasses = {
  solid:
    'text-white disabled:border-gray-100 disabled:text-gray-400 disabled:bg-gray-100',
  outlined:
    'bg-opacity-0 hover:bg-opacity-20 disabled:border-gray-300 disabled:text-gray-300',
  text: 'disabled:text-gray-300 border-transparent bg-transparent',
};
const solidButtonVariantClasses = {
  primary:
    'border-primary bg-primary hover:bg-primary-dark hover:border-primary-dark',
  secondary:
    'border-secondary bg-secondary hover:bg-secondary-dark hover:border-secondary-dark',
  error: 'border-error bg-error hover:bg-error-dark hover:border-error-dark',
  warning:
    'border-warning bg-warning hover:bg-warning-dark hover:border-warning-dark',
  info: 'border-info bg-info hover:bg-info-dark hover:border-info-dark',
  success:
    'border-success bg-success hover:bg-success-dark hover:border-success-dark',
};
const outlinedButtonVariantClasses = {
  primary: 'text-primary border-primary bg-primary',
  secondary: 'text-secondary border-secondary bg-secondary',
  error: 'text-error border-error bg-error',
  warning: 'text-warning border-warning bg-warning',
  info: 'text-info border-info bg-info',
  success: 'text-success border-success bg-success',
};
const textButtonVariantClasses = {
  primary: 'text-primary hover:text-primary-dark',
  secondary: 'text-secondary hover:text-secondary-dark',
  error: 'text-error hover:text-error-dark',
  warning: 'text-warning hover:text-warning-dark',
  info: 'text-info hover:text-info-dark',
  success: 'text-success hover:text-success-dark',
};

function BaseButton({
  size,
  buttonType,
  variant,
  rounded,
  children,
}: BaseButtonProps) {
  return (
    <button
      className={cn(
        'inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold leading-none duration-200',
        'disabled:pointer-events-none disabled:cursor-default',
        buttonSizeClasses[size],
        buttonType === 'solid' && solidButtonVariantClasses[variant],
        buttonType === 'outlined' && outlinedButtonVariantClasses[variant],
        buttonType === 'text' && textButtonVariantClasses[variant],
        buttonTypeClasses[buttonType],
        !rounded && 'rounded-none'
      )}
    >
      {children}
    </button>
  );
}

function Icon({ size, children }: IconProps) {
  return (
    <div
      className={cn(
        size === Size.small && 'h-5 w-5',
        size === Size.medium && 'h-6 w-6',
        size === Size.large && 'h-7 w-7'
      )}
    >
      {children}
    </div>
  );
}

export function Button({
  testId = 'nxact-ui-button',
  size = Size.medium,
  buttonType = 'solid',
  variant = ColorVariant.primary,
  rounded = true,
  leftIcon,
  rightIcon,
  children,
  ...props
}: ButtonProps) {
  return (
    <BaseButton
      data-testid={testId}
      size={size}
      buttonType={buttonType}
      variant={variant}
      rounded={rounded}
      {...props}
    >
      {!!leftIcon && (
        <Icon size={size}>
          {cloneElement(leftIcon as React.ReactElement, {
            width: '100%',
            height: '100%',
          })}
        </Icon>
      )}
      {children}
      {!!rightIcon && (
        <Icon size={size}>
          {cloneElement(rightIcon as React.ReactElement, {
            width: '100%',
            height: '100%',
          })}
        </Icon>
      )}
    </BaseButton>
  );
}
