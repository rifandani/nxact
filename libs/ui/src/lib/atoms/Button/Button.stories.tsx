import type { ComponentMeta, ComponentStoryFn } from '@storybook/react';
import { ColorVariant, Size } from '../../../types/common.type';
import { Button } from './Button';

const Story: ComponentMeta<typeof Button> = {
  title: 'atoms/Button',
  component: Button,
  argTypes: {
    onClick: {
      action: 'clicked',
    },
    size: {
      control: { type: 'select', options: Size },
    },
    buttonType: {
      control: { type: 'select', options: ['solid', 'outlined', 'text'] },
    },
    variant: {
      control: { type: 'select', options: ColorVariant },
    },
    rounded: {
      control: { type: 'boolean' },
    },
    disabled: {
      control: { type: 'boolean' },
    },
    leftIcon: {
      control: 'boolean',
    },
    rightIcon: {
      control: 'boolean',
    },
  },
};
export default Story;

const Template: ComponentStoryFn<typeof Button> = (args) => (
  <Button {...args} />
);

export const Default = Template.bind({});
export const Outlined = Template.bind({});
export const Text = Template.bind({});

Default.args = {
  buttonType: 'solid',
  children: 'Solid',
};
Outlined.args = {
  buttonType: 'outlined',
  children: 'Outlined',
};
Text.args = {
  buttonType: 'text',
  children: 'Text',
};
