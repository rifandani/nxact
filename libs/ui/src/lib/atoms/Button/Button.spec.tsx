import { Button } from 'libs/ui/src/lib/atoms/Button/Button';
import { Size } from 'libs/ui/src/types/button.type';
import renderer from 'react-test-renderer';

describe('Button regular component', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<Button>Button Text</Button>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render correctly when there's no child component", () => {
    const tree = renderer.create(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with text child and type', () => {
    const tree = renderer
      .create(<Button type="submit">Button Text</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when component is disabled', () => {
    const tree = renderer
      .create(<Button disabled>Button Text</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should be able to fire onClick event', () => {
    const onClick = jest.fn();
    const tree = renderer.create(<Button onClick={onClick} />);
    tree.root.props.onClick();
    expect(onClick).toHaveBeenCalled();
  });

  it('should render correctly with small prop', () => {
    const tree = renderer
      .create(<Button size={Size.small}>Button Text</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with no rounded prop', () => {
    const tree = renderer.create(<Button>Button Text</Button>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button Outlined component', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(<Button buttonType="outlined">Button Text</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render correctly when there's no child component", () => {
    const tree = renderer.create(<Button buttonType="outlined" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with text child and type', () => {
    const tree = renderer
      .create(
        <Button buttonType="outlined" type="submit">
          Button Text
        </Button>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when component is disabled', () => {
    const tree = renderer
      .create(
        <Button buttonType="outlined" disabled>
          Button Text
        </Button>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should be able to fire onClick event', () => {
    const onClick = jest.fn();
    const tree = renderer.create(
      <Button buttonType="outlined" onClick={onClick} />
    );
    tree.root.props.onClick();
    expect(onClick).toHaveBeenCalled();
  });

  it('should render correctly with small prop', () => {
    const tree = renderer
      .create(
        <Button buttonType="outlined" size={Size.small}>
          Button Text
        </Button>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with no rounded prop', () => {
    const tree = renderer
      .create(<Button buttonType="outlined">Button Text</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button Text component', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(<Button buttonType="outlined">Button Text</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render correctly when there's no child component", () => {
    const tree = renderer.create(<Button buttonType="text" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with text child and type', () => {
    const tree = renderer
      .create(
        <Button buttonType="text" type="submit">
          Button Text
        </Button>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when component is disabled', () => {
    const tree = renderer
      .create(
        <Button buttonType="text" disabled>
          Button Text
        </Button>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should be able to fire onClick event', () => {
    const onClick = jest.fn();
    const tree = renderer.create(
      <Button buttonType="text" onClick={onClick} />
    );
    tree.root.props.onClick();
    expect(onClick).toHaveBeenCalled();
  });

  it('should render correctly with small prop', () => {
    const tree = renderer
      .create(
        <Button buttonType="text" size={Size.small}>
          Button Text
        </Button>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
