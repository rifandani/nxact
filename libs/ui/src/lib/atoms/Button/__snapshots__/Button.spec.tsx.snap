// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Outlined component should render correctly 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md text-primary border-primary bg-primary bg-opacity-0 hover:bg-opacity-20 disabled:border-gray-300 disabled:text-gray-300"
  data-testid="nxact-ui-button"
>
  Button Text
</button>
`;

exports[`Button Outlined component should render correctly when component is disabled 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md text-primary border-primary bg-primary bg-opacity-0 hover:bg-opacity-20 disabled:border-gray-300 disabled:text-gray-300"
  data-testid="nxact-ui-button"
  disabled={true}
>
  Button Text
</button>
`;

exports[`Button Outlined component should render correctly when there's no child component 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md text-primary border-primary bg-primary bg-opacity-0 hover:bg-opacity-20 disabled:border-gray-300 disabled:text-gray-300"
  data-testid="nxact-ui-button"
/>
`;

exports[`Button Outlined component should render correctly with no rounded prop 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md text-primary border-primary bg-primary bg-opacity-0 hover:bg-opacity-20 disabled:border-gray-300 disabled:text-gray-300"
  data-testid="nxact-ui-button"
>
  Button Text
</button>
`;

exports[`Button Outlined component should render correctly with small prop 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-1.5 text-sm rounded text-primary border-primary bg-primary bg-opacity-0 hover:bg-opacity-20 disabled:border-gray-300 disabled:text-gray-300"
  data-testid="nxact-ui-button"
>
  Button Text
</button>
`;

exports[`Button Outlined component should render correctly with text child and type 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md text-primary border-primary bg-primary bg-opacity-0 hover:bg-opacity-20 disabled:border-gray-300 disabled:text-gray-300"
  data-testid="nxact-ui-button"
  type="submit"
>
  Button Text
</button>
`;

exports[`Button Text component should render correctly 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md text-primary border-primary bg-primary bg-opacity-0 hover:bg-opacity-20 disabled:border-gray-300 disabled:text-gray-300"
  data-testid="nxact-ui-button"
>
  Button Text
</button>
`;

exports[`Button Text component should render correctly when component is disabled 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md text-primary hover:text-primary-dark disabled:text-gray-300 border-transparent bg-transparent"
  data-testid="nxact-ui-button"
  disabled={true}
>
  Button Text
</button>
`;

exports[`Button Text component should render correctly when there's no child component 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md text-primary hover:text-primary-dark disabled:text-gray-300 border-transparent bg-transparent"
  data-testid="nxact-ui-button"
/>
`;

exports[`Button Text component should render correctly with small prop 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-1.5 text-sm rounded text-primary hover:text-primary-dark disabled:text-gray-300 border-transparent bg-transparent"
  data-testid="nxact-ui-button"
>
  Button Text
</button>
`;

exports[`Button Text component should render correctly with text child and type 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md text-primary hover:text-primary-dark disabled:text-gray-300 border-transparent bg-transparent"
  data-testid="nxact-ui-button"
  type="submit"
>
  Button Text
</button>
`;

exports[`Button regular component should render correctly 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md border-primary bg-primary hover:bg-primary-dark hover:border-primary-dark text-white disabled:border-gray-100 disabled:text-gray-400 disabled:bg-gray-100"
  data-testid="nxact-ui-button"
>
  Button Text
</button>
`;

exports[`Button regular component should render correctly when component is disabled 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md border-primary bg-primary hover:bg-primary-dark hover:border-primary-dark text-white disabled:border-gray-100 disabled:text-gray-400 disabled:bg-gray-100"
  data-testid="nxact-ui-button"
  disabled={true}
>
  Button Text
</button>
`;

exports[`Button regular component should render correctly when there's no child component 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md border-primary bg-primary hover:bg-primary-dark hover:border-primary-dark text-white disabled:border-gray-100 disabled:text-gray-400 disabled:bg-gray-100"
  data-testid="nxact-ui-button"
/>
`;

exports[`Button regular component should render correctly with no rounded prop 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md border-primary bg-primary hover:bg-primary-dark hover:border-primary-dark text-white disabled:border-gray-100 disabled:text-gray-400 disabled:bg-gray-100"
  data-testid="nxact-ui-button"
>
  Button Text
</button>
`;

exports[`Button regular component should render correctly with small prop 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-1.5 text-sm rounded border-primary bg-primary hover:bg-primary-dark hover:border-primary-dark text-white disabled:border-gray-100 disabled:text-gray-400 disabled:bg-gray-100"
  data-testid="nxact-ui-button"
>
  Button Text
</button>
`;

exports[`Button regular component should render correctly with text child and type 1`] = `
<button
  className="inline-flex cursor-pointer items-center justify-center space-x-3 border border-solid px-4 font-semibold duration-200 disabled:pointer-events-none disabled:cursor-default py-[9px] text-base rounded-md border-primary bg-primary hover:bg-primary-dark hover:border-primary-dark text-white disabled:border-gray-100 disabled:text-gray-400 disabled:bg-gray-100"
  data-testid="nxact-ui-button"
  type="submit"
>
  Button Text
</button>
`;
